t/plugin/file-logger2.t

```diff
@@ -0,0 +1,186 @@
+            msg = fd:read()
+
+            local new_msg = core.json.decode(msg)
+            ngx.status = code
+
+            if new_msg.response ~=nil and new_msg.response.body == "hello world\n"
+            then
+                ngx.status = code
+                ngx.say('contain with target')
+            end
+        }
+    }
+--- response_body
+contain with target
+
+
+
+=== TEST 3: check file-logger 'include_resp_body' with 'expr'
+--- config
+    location /t {
+        content_by_lua_block {
+            local t = require("lib.test_admin").test
+            local code, body = t('/apisix/admin/routes/1',
+                 ngx.HTTP_PUT,
+                 [[{
+                        "plugins": {
+                            "file-logger": {
+                                "path": "file-with-resp-expr-body.log",
+                                "include_resp_body": true,
+                                "include_resp_body_expr": [
+                                    [
+                                      "arg_foo",
+                                      "==",
+                                      "bar"
+                                    ]
+                                ]
+                            }
+                        },
+                        "upstream": {
+                            "nodes": {
+                                "127.0.0.1:1982": 1
+                            },
+                            "type": "roundrobin"
+                        },
+                        "uri": "/hello"
+                }]]
+                )
+
+            if code >= 300 then
+                ngx.status = code
+            end
+            ngx.say(body)
+        }
+    }
+
+
+
+=== TEST 4: verify file-logger resp with expression of concern
+--- config
+    location /t {
+        content_by_lua_block {
+            local core = require("apisix.core")
+            local t = require("lib.test_admin").test
+            local code = t("/hello?foo=bar", ngx.HTTP_GET)
+            local fd, err = io.open("file-with-resp-expr-body.log", 'r')
+            local msg
+
+            if not fd then
+                core.log.error("failed to open file: file-with-resp-expr-body.log, error info: ", err)
+                return
+            end
+
+            -- note only for first line
+            msg = fd:read()
+
+            local new_msg = core.json.decode(msg)
+            ngx.status = code
+
+            if new_msg.response ~=nil and new_msg.response.body == "hello world\n"
+            then
+                ngx.status = code
+                ngx.say('contain target body hits with expr')
+            end
+
+            --- a new request is logged
+            t("/hello?name=pix", ngx.HTTP_GET)
+            msg = fd:read("*l")
+            local new_msg = core.json.decode(msg)
+            if new_msg.response.body == nil
+            then
+                ngx.say('skip unconcern body')
+            end
```

Comment:
here the `msg` is read by `msg = fd:read()` before the new request is sent, so it's not possible for it to record new requests about it?

I suggest writing the new request as a new test case.

Reply:
Thanks for your follow, here's the case:
Request-1, `fd:read()` read the first line content,
Request-2, `fd:read("*l")` read the next line.
What the log file generates:
```
1 {"upstream":"127.0.0.1:1982","service_id":"","server":{"hostname":"pixelpig","version":"3.0.0"},"request":{"querystring":{"foo":"bar"},"headers":{"content-type":"application\/x-www-form-urlencoded","user-agent":"lua-resty-http\/0.16.1 (Lua) ngx_lua\/10021","host":"127.0.0.1:1984"},"method":"GET","size":157,"uri":"\/hello?foo=bar","url":"http:\/\/127.0.0.1:1984\/hello?foo=bar"},"apisix_latency":100.99994659424,"start_time":1670294090157,"route_id":"1","upstream_latency":5,"latency":105.99994659424,"response":{"size":123,"headers":{"server":"APISIX\/3.0.0","content-type":"text\/plain","content-length":"12","connection":"close"},"body":"hello world\n","status":200},"client_ip":"127.0.0.1"}
2 {"upstream":"127.0.0.1:1982","service_id":"","server":{"hostname":"pixelpig","version":"3.0.0"},"request":{"querystring":{"name":"pix"},"headers":{"content-type":"application\/x-www-form-urlencoded","user-agent":"lua-resty-http\/0.16.1 (Lua) ngx_lua\/10021","host":"127.0.0.1:1984"},"method":"GET","size":158,"uri":"\/hello?name=pix","url":"http:\/\/127.0.0.1:1984\/hello?name=pix"},"apisix_latency":100.00016403198,"start_time":1670294090274,"route_id":"1","upstream_latency":3,"latency":103.00016403198,"response":{"headers":{"server":"APISIX\/3.0.0","content-type":"text\/plain","content-length":"12","connection":"close"},"size":123,"status":200},"client_ip":"127.0.0.1"}
```
With the read cursor moving, that's why the branch `if new_msg.response.body == nil` will hit, Is my understanding right?

Reply:
yes, my mistake.